### PR TITLE
Update backend workflow such that frontend can run in full and door unlock testing can be done

### PR DESCRIPTION
**Pull request by ABarroso647**
Changes include:
-Updating association to be user->roles
-Updating user post request to allow the addition of roles on creation example request shown below:

-Updating the user put request to allow the updating of both roles and user fields at the same time with one reuqest, example request shown below:

-Adding getAllkeys function to get all keys for a user

-Adding delete propogation such that, when deleting associations are automatically deleted and in the case of many to one, like our users, all associated keys are deleted upon user delete

-Updated role post similarily to users to allow door associations to be added on create.

-Adding in full workflow to check if user has access to the door upon login, and then to unlock door (mobile flow),
-Bugfix on login logic

### CODE REVIEW

File: backend/config/db.go [hunk 1]
```diff
@@ -15,7 +16,7 @@ func Connect() {
 	if err != nil {
 		panic("failed to create data directory")
 	}
-	db, err := gorm.Open(sqlite.Open("data/lynx-web.db"), &gorm.Config{TranslateError: true})
+	db, err := gorm.Open(sqlite.Open("data/lynx-web.db"), &gorm.Config{TranslateError: true, Logger: logger.Default.LogMode(logger.Info)})
```
Comment (AdrianJendo) [inline backend/config/db.go]: what for?

File: backend/controller/role.go [hunk 2]
```diff
@@ -41,6 +41,14 @@ func CreateRole(w http.ResponseWriter, r *http.Request) {
...
 		http.Error(w, "Malformed request", http.StatusBadRequest)
 	}
+	doors := role.Doors
+	for i, _ := range doors {
+		res := db.DB.First(&doors[i])
+		if res.Error != nil {
+			http.Error(w, "One or more invalid doors entered", http.StatusBadRequest)
+			return
+		}
+	}
```
Comment (AdrianJendo) [inline backend/controller/role.go]: This seems wildly inefficient, can we do something like this and compare the lengths:

Screenshot 2024-03-11 at 3 38 30 PM
Screenshot 2024-03-11 at 3 39 10 PM
Screenshot 2024-03-11 at 3 39 01 PM
Screenshot 2024-03-11 at 3 38 50 PM

Reference: https://gorm.io/docs/query.html#Retrieving-all-objects

File: backend/controller/user.go [hunk 3]
```diff
@@ -34,15 +34,27 @@ func UpdateUser(w http.ResponseWriter, r *http.Request) {
...
 	}
+	roles := user.Roles
+	for i, _ := range roles {
+		res := db.DB.First(&roles[i])
+		if res.Error != nil {
+			http.Error(w, "One or more invalid roles entered", http.StatusBadRequest)
+			return
+		}
+	}
 	err, user = helpers.UpdateObject(w, user)
```
Comment (AdrianJendo) [inline backend/controller/user.go]: Same here, let's not do n separate SELECT statements when we can just do 1.

File: backend/controller/user.go [hunk 4]
```diff
@@ -34,15 +34,27 @@ func UpdateUser(w http.ResponseWriter, r *http.Request) {
...
 	if err != nil {
 		return
 	}
+	err = db.DB.Debug().Model(&user).Association("Roles").Replace(&roles)
```
Comment (AdrianJendo) [inline backend/controller/user.go]: .Debug

File: backend/controller/user.go [hunk 5]
```diff
@@ -67,9 +79,18 @@ func GetUserByEmail(w http.ResponseWriter, r *http.Request) {
...
+
 	err := json.NewDecoder(r.Body).Decode(&user)
+	roles := user.Roles
+	for i, _ := range roles {
+		res := db.DB.First(&roles[i])
+		if res.Error != nil {
+			http.Error(w, "One or more invalid roles entered", http.StatusBadRequest)
+			return
+		}
+	}
```
Comment (AdrianJendo) [inline backend/controller/user.go]: same here

File: backend/controller/webAuthn.go [hunk 6]
```diff
@@ -155,8 +156,7 @@ func RegisterResponse(w http.ResponseWriter, r *http.Request) {
 	// Add public key to DB
 	dbCredential := models.Credential{
 		UserId:          user.Id,
-		Roles:           nil, // TODO: use activeTokens table to find roles
-		WebauthnId:      credential.ID,
+		Id:              credential.ID,
```
Comment (AdrianJendo) [inline backend/controller/webAuthn.go:159]: Let's move this to the first row in the struct.

File: backend/controller/webAuthn.go [hunk 7]
```diff
@@ -266,15 +268,44 @@ func AuthorizeResponse(w http.ResponseWriter, r *http.Request) {
...
 		return
 	}
-
-	err = SaveDBCredential(w, credential)
-	if err != nil {
+	//err = SaveDBCredential(w, credential)
+	//if err != nil {
+	//	http.Error(w, "Could not finish log in", http.StatusInternalServerError)
+	//	return
+	//}
```
Comment (AdrianJendo) [inline backend/controller/webAuthn.go:280]: Let's add a comment explaining what this is and why it was commented out.

File: backend/controller/webAuthn.go [hunk 8]
```diff
@@ -266,15 +268,44 @@ func AuthorizeResponse(w http.ResponseWriter, r *http.Request) {
...
+
+	user := models.User{}
+	creds := models.Credential{Id: credential.ID}
+	res := db.DB.First(&creds)
+	if res.Error != nil {
+		http.Error(w, "Could retrieve credentials ", http.StatusInternalServerError)
```
Comment (AdrianJendo) [inline backend/controller/webAuthn.go]: "Could not retrieve credential"

File: backend/controller/webAuthn.go [hunk 9]
```diff
@@ -266,15 +268,44 @@ func AuthorizeResponse(w http.ResponseWriter, r *http.Request) {
...
+	db.DB.Model(&creds).Association("User").Find(&user)
+	roles := []models.Role{}
+	db.DB.Model(&user).Association("Roles").Find(&roles)
+	doors := []models.Door{}
+	db.DB.Distinct("id").Model(&roles).Association("Doors").Find(&doors)
+	hasAccess := false
+	for _, curDoor := range doors {
+		if curDoor.Id == dId {
+			hasAccess = true
+		}
```
Comment (AdrianJendo) [inline backend/controller/webAuthn.go]: Let's break if we find the id

File: backend/controller/webAuthn.go [hunk 10]
```diff
@@ -266,15 +268,44 @@ func AuthorizeResponse(w http.ResponseWriter, r *http.Request) {
...
+	}
+	if hasAccess == false {
```
Comment (AdrianJendo) [inline backend/controller/webAuthn.go]: `!hasAccess`

File: backend/controller/webAuthn.go (hunk 9 above)
Comment (AdrianJendo) [reply backend/controller/webAuthn.go]: Actually, can we just use a built-in search function: https://stackoverflow.com/questions/38654383/how-to-search-for-an-element-in-a-golang-slice
https://pkg.go.dev/slices#IndexFunc

File: backend/controller/webAuthn.go [hunk 11]
```diff
@@ -266,15 +268,44 @@ func AuthorizeResponse(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, "User does not have access to door", http.StatusBadRequest)
+		return
+	}
+	errJson := json.NewEncoder(w).Encode("Login Success, opening door")
 	if errJson != nil {
 		http.Error(w, "Could not return results", http.StatusInternalServerError)
 		return
 	}
+	go unlockDoor(dId)
```
Comment (AdrianJendo) [inline backend/controller/webAuthn.go]: Why do you write `go unlockDoor()` instead of `unlockDoor()`

Comment (AdrianJendo) [inline backend/controller/webAuthn.go]:
```
x := slices.IndexFunc(doors, func(d modes.Door) bool { return d.Id == dId }
if x == -1 {
  // error logic
  return
}

// success logic

```

File: backend/routes/door.go [hunk 12]
```diff
@@ -11,5 +11,6 @@ func DoorsRoute() chi.Router {
 	r.Post("/", controller.CreateDoor)
 	r.Put("/", controller.UpdateDoor)
 	r.Delete("/{doorId}", controller.DeleteDoor)
+	r.Get("/unlocked/{doorId}", controller.OpenDoor)
```
Comment (AdrianJendo) [inline backend/routes/door.go]: `unlocked` or `unlock`?

File: backend/routes/user.go [hunk 13]
```diff
@@ -13,7 +13,7 @@ func UsersRoute() chi.Router {
 	r.Put("/", controller.UpdateUser)
 	r.Post("/", controller.CreateUser)
 	r.Delete("/{userId}", controller.DeleteUser)
-	r.Post("/{userId}/roles", controller.ReplaceRoleAssociations)
 	r.Get("/{userId}/roles", controller.GetRoleAssociations)
+	r.Get("/{userId}/creds", controller.GetCredAssociations)
```
Comment (AdrianJendo) [inline backend/routes/user.go]: Can we call these like `GetUserRoles` and `GetUserCreds` or something

**Review by AdrianJendo — COMMENTED** [review]
I don't think this should all be in 1 PR for the future

File: backend/controller/webAuthn.go (hunk 11 above)
Comment (ABarroso647) [reply backend/controller/webAuthn.go]: what does this refer to?

File: backend/routes/door.go (hunk 12 above)
Comment (ABarroso647) [reply backend/routes/door.go]: well a door is "unlocked" not unlock

Comment (AdrianJendo) [reply backend/routes/door.go]: What does this endpoint do?

Comment (ABarroso647) [reply backend/routes/door.go]: this is the endpoint cameron is pinging, so it tells him when a door has been "unlocked" by a user so he can run the logic to open it

Comment (AdrianJendo) [reply backend/routes/door.go]: Can you change the controller name this to `IsDoorOpen` or something

**Review by AdrianJendo — APPROVED** [review]

**Pull request merged**